# II deo
### Pre same obrade podataka i donošenja zaključaka, neophodno je tzv. "sirovi" (raw) dataset da sredimo tako da bude pogodan za obradu. Jedna od glavnih aktivnosti u okviru tog sređivanja jeste rešavanje problema nedostajućih vrednosti.

In [2]:
import pandas as pd
from pandas import Series

dfr = pd.read_csv('../data/brit.csv')
dfr.sample(10)

Title                   AlbumName  \
1122      Love Is All Around                   Cellophane   
59              I Meant You            Satisfied with You   
992                Each Time   Take Me for What I'm Worth   
431          Pink Champagne                  Fame at Last   
1037                   Today      Love Me, Please Love Me   
1217  I Got My Mojo Working                    Begin Here   
432         Monkeying Around                 Fame at Last   
1117            It's Showing                   Cellophane   
761     A Well Respected Man                  Kinda Kinks   
338         The War Drags On                    Fairytale   

                                           Record Label  \
1122                                         Repertoire   
59                                                  BMG   
992                                                 Pye   
431                                 Polydor / Universal   
1037                                                Pye   
1217                                   Big Beat Records   
432                                 Polydor / Universal   
1117                                         Repertoire   
761                                           Sanctuary   
338   Castle Music Ltd. / Fontana / Sanctuary / Sanc...   

                                              Performer  Track.number  \
1122                                         The Troggs            12   
59                                  The Dave Clark Five             4   
992                                       The Searchers             9   
431   Georgie Fame                                  ...             7   
1037                                        Sandie Shaw            17   
1217                                        The Zombies            14   
432                                        Georgie Fame             8   
1117                                         The Troggs             7   
761                                           The Kinks            21   
338                                             Donovan            15   

     Song.duration                                     Composer  Year  \
1122          3:00                                  Reg Presley  1967   
59            1:59                    Dave Clark / Denis Payton  1966   
992           2:48                             Jackie DeShannon  1965   
431           3:54  Harold Adamson / Leo McCarey / Harry Warren  1965   
1037          2:24                                Chris Andrews  1967   
1217          3:36        Preston Foster / McKinley Morganfield  1965   
432           2:12                                 William Bell  1965   
1117          2:56                                  Reg Presley  1967   
761           2:43                                   Ray Davies  1965   
338           3:44                                 Mick Softley  1965   

                           Genre  \
1122                    Pop/Rock   
59                      Pop/Rock   
992                     Pop/Rock   
431   Pop/Rock, R&B, Jazz, Blues   
1037                    Pop/Rock   
1217                    Pop/Rock   
432   Pop/Rock, R&B, Jazz, Blues   
1117                    Pop/Rock   
761                     Pop/Rock   
338                         Folk   

                                                 Styles  ...  speechiness  \
1122    British Invasion, AM Pop, Contemporary Pop/Rock  ...       0.0273   
59                        British Invasion, Rock & Roll  ...       0.0302   
992   British Invasion, Contemporary Pop/Rock, Folk-...  ...       0.0418   
431   British Invasion, Blue-Eyed Soul, Soul, Vocal ...  ...       0.0279   
1037                           British Invasion, AM Pop  ...       0.0297   
1217  British Invasion, Baroque Pop, AM Pop, Contemp...  ...       0.0952   
432   British Invasion, Blue-Eyed Soul, Soul, Vocal ...  ...       0.0482   
1117    British Invasion, AM Pop, Contemporary Pop/Rock  ...          NaN   
761 

### Najpre ćemo utvrditi koliko nedostajućih (NaN/None) vrednosti ima u svakoj koloni. Metoda isna() preslikava svako polje datafrejma u True/False (ako je polje NaN dobiće vrednost False).

In [2]:
dfr.isna().sum().sort_values(ascending=False)

duration_ms           127
valence               127
tempo                 127
speechiness           127
loudness              127
liveness              127
instrumentalness      127
energy                127
danceability          127
acousticness          127
popularity            127
Song.duration          43
Record Label            3
Composer                0
Track.number            0
Performer               0
Title                   0
AlbumName               0
Styles                  0
Genre                   0
Year                    0
shake.the.audience      0
romantic                0
obscene                 0
family.spiritual        0
sadness                 0
dtype: int64

## 1. varijanta: popunjavanje nedostajućih na osnovu poznatih vrednosti
 ### Nedostajućim vrednostima kolone `loudness` biće dodeljena srednja vrednost te kolone. Nakon popunjavanja nedostajucih vrednosti korišćenjem metode `fillna`, metoda `value_counts` nam vraća broj pojavljivanja svake jedinstvene vrednosti. Srednja vrednost se pojavljuje najčešće, i to kod svih 127 prethodno nedostajućih polja.

In [3]:
dfr['loudness'] = dfr['loudness'].fillna(value=dfr['loudness'].mean())
dfr['loudness'].value_counts()

loudness
-9.566823     127
-9.197000       3
-8.887000       3
-9.436000       3
-10.086000      3
             ... 
-9.310000       1
-9.562000       1
-9.299000       1
-11.992000      1
-9.279000       1
Name: count, Length: 1028, dtype: int64

## 2. varijanta: uklanjanje redova koji sadrže nedostajuće vrednosti
 ### Uklonićemo sve redove kod kojih kolonama energy i tempo nedostaje vrednost, koristeći metodu `dropna`. Kod `subset` parametra navode se kolone za koje se nedostajuće vrednosti razmatraju, dok parametar `how` uzima vrednost `all` (red se briše samo ako su sve vrednosti kolona iz subseta nedostajuće) ili `any` (red se briše ako nedostaje bilo koja vrednost iz subseta). Kako bismo efekat ove naredbe preneli na naš datafrejm, parametar `inplace` postavljamo na True.

In [25]:
dfr.dropna(subset=['energy','tempo'],how='all', inplace=True)
dfr.isna().sum().sort_values(ascending=False)

Song.duration         32
Record Label           3
AlbumName              0
Title                  0
Performer              0
Track.number           0
Composer               0
Year                   0
Genre                  0
Styles                 0
acousticness           0
danceability           0
energy                 0
instrumentalness       0
liveness               0
loudness               0
speechiness            0
tempo                  0
valence                0
duration_ms            0
popularity             0
shake.the.audience     0
romantic               0
obscene                0
family.spiritual       0
sadness                0
dtype: int64

### Ovakav način regulisanja nedostajućih vrednosti pogodniji je za nenumeričke kolone, jer kod njih nije moguće dobiti srednju vrednost kolone. Kolona `Record Label` je kategorička, stoga je kod nje najbolje primeniti metodu `dropna`.

In [38]:
dfr.dropna(subset=['Record Label', 'Song.duration'], inplace=True)

Title                 0
AlbumName             0
Record Label          0
Performer             0
Track.number          0
Song.duration         0
Composer              0
Year                  0
Genre                 0
Styles                0
acousticness          0
danceability          0
energy                0
instrumentalness      0
liveness              0
loudness              0
speechiness           0
tempo                 0
valence               0
duration_ms           0
popularity            0
shake.the.audience    0
romantic              0
obscene               0
family.spiritual      0
sadness               0
dtype: int64

### Budući da smo otklonili sve nedostajuće vrednosti, izmenjeni dataset ćemo sačuvati u novi csv fajl.

In [37]:
dfr.to_csv('../data/brit_bez_NaN.csv', index=False)

### Sada ćemo učitati fajl koji smo sačuvali kako bismo nastavili njegovo sređivanje. Izvršićemo sređivanje tekstualnih kolona. U datafrejmu ove kolone su tipa `object`.

In [40]:
dfr2 = pd.read_csv('../data/brit_bez_NaN.csv')
dfr2.select_dtypes(include='object')

Title      AlbumName               Record Label  \
0                 Glad All Over    Glad All Over                        BMG   
1               All of the Time    Glad All Over                        BMG   
2               Crying Over You    Glad All Over                        BMG   
3                          Stay    Glad All Over                        BMG   
4                      Chaquita    Glad All Over                        BMG   
...                           ...            ...                        ...   
1071                 Leave Me Be      I Love You  Varèse Sarabande / Varese   
1072     I Can't Make up My Mind      I Love You  Varèse Sarabande / Varese   
1073                Remember You      I Love You  Varèse Sarabande / Varese   
1074           Just Out of Reach      I Love You  Varèse Sarabande / Varese   
1075  She Does Everything for Me      I Love You  Varèse Sarabande / Varese   

                Performer Song.duration                         Composer  \
0     The Dave Clark Five          2:43          Dave Clark / Mike Smith   
1     The Dave Clark Five          2:15          Dave Clark / Mike Smith   
2     The Dave Clark Five          2:09      Dave Clark / Lenny Davidson   
3     The Dave Clark Five          2:09  Sobian Fahey / Maurice Williams   
4     The Dave Clark Five          2:13          Dave Clark / Mike Smith   
...                   ...           ...                              ...   
1071          The Zombies          2:08                      Chris White   
1072          The Zombies          2:33                      Chris White   
1073          The Zombies          1:58                      Chris White   
1074          The Zombies          2:08                  Colin Blunstone   
1075          The Zombies          2:15                       Rod Argent   

         Genre                                             Styles  
0     Pop/Rock      British Invasion, Early Pop/Rock, Rock & Roll  
1     Pop/Rock      British Invasion, Early Pop/Rock, Rock & Roll  
2     Pop/Rock      British Invasion, Early Pop/Rock, Rock & Roll  
3     Pop/Rock      British Invasion, Early Pop/Rock, Rock & Roll  
4     Pop/Rock      British Invasion, Early Pop/Rock, Rock & Roll  
...        ...                                                ...  
1071  Pop/Rock  British Invasion, Baroque Pop, AM Pop, Contemp...  
1072  Pop/Rock  British Invasion, Baroque Pop, AM Pop, Contemp...  
1073  Pop/Rock  British Invasion, Baroque Pop, AM Pop, Contemp...  
1074  Pop/Rock  British Invasion, Baroque Pop, AM Pop, Contemp...  
1075  Pop/Rock  British Invasion, Baroque Pop, AM Pop, Contemp...  

[1076 rows x 8 columns]

### Iz kolone `Composer` ćemo izdvojiti prvog kompozitora i sačuvati ga u novu kolonu `First composer`

In [30]:
def izdvoj_prvog_komp(comps):
    kl =  comps.split('/')
    return kl[0].strip()
print(izdvoj_prvog_komp('Dave Clark / Mike Smith'))

Dave Clark


### Sada ćemo ovu funkciju primeniti (funkcija `apply`) na svako polje iz kolone `Composer`, i na taj način ćemo dobiti željenu novu kolonu.

In [33]:
dfr2['First composer'] = dfr2['Composer'].apply(izdvoj_prvog_komp)
dfr2['First composer'].sample(10)

1071      Jim McCarty
607      Allan Clarke
844       Mick Jagger
296           Donovan
957     Jerry Ragovoy
398          Fred Jay
688        Ray Davies
95        Fats Domino
756          Phil May
273           Ed Cobb
Name: First composer, dtype: object

### Kod kolone `Performer` postoji problem da se umetnuo znak za novi red (\n) kod određenih izvođača. Rešićemo to preko funkcije strip (split u javi). 

In [41]:
dfr2['Performer'].str.contains('\n').sum()

np.int64(47)

### Primećujemo da ima 47 zapisa koji sadrže znak ove za novi red. Ovaj put nećemo pisati funkciju za otklon tih znakova pre samog poziva apply funkcije, već ćemo to uraditi direktno u njoj preko lambda izraza.

In [42]:
dfr2['Performer'] = dfr2['Performer'].apply(lambda izv: izv.replace('\n', ''))

dfr2['Performer'].str.contains('\n').sum()

np.int64(0)

### Nakon što smo rešili ovaj problem, utvrdili smo da je sem znakova za novi red, umetnuto pedesetak znakova za razmak u samom nazivu izvođača.

In [44]:
def brisi_razmake(izv):
    reci = izv.split(' ')
    words = [r for r in reci if len(r)>0]
    return ' '.join(words)

dfr2['Performer'] = dfr2['Performer'].apply(brisi_razmake)

### Nakon što smo rešili i ovaj problem, kolona Performers je ispravna, stoga možemo sačuvati promene u novi csv fajl.

In [45]:
dfr2.to_csv('../data/brit_sredjen.csv', index=False)